### PR TITLE
refactor(components): [progress] switch to script-setup syntax

### PR DESCRIPTION
**Pull request by chenxch**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

- switch to script-setup syntax #6114 
- Add `ProgressInstance` type

### CODE REVIEW

File: packages/components/progress/src/progress.vue [hunk 1]
```diff
@@ -88,176 +88,146 @@ import {
 } from '@element-plus/icons-vue'
 import { useNamespace } from '@element-plus/hooks'
 import { progressProps } from './progress'
+import type { Color } from './progress'
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:91]: **Suggestion:** put line#91 after line#92

File: packages/components/progress/src/progress.vue [hunk 2]
```diff
@@ -88,176 +88,146 @@ import {
...
+    backgroundColor: getCurrentColor(props.percentage),
+  })
+)
+
+const relativeStrokeWidth = computed(() =>
+  ((props.strokeWidth / props.width) * 100).toFixed(1)
+)
+
+const radius = computed(() => {
+  if (props.type === 'circle' || props.type === 'dashboard') {
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:122]: **Suggestion:** 

```suggestion
  if (['circle', 'dashboard'].includes(props.type)) {
```

File: packages/components/progress/src/progress.vue [hunk 3]
```diff
@@ -88,176 +88,146 @@ import {
...
-        return colors[colors.length - 1]?.color
-      }
-    }
+const content = computed(() => props.format(props.percentage))
 
-    const slotData = computed(() => {
+function getColors(color: Color[]) {
+  const span = 100 / color.length
+  const seriesColors = color.map((seriesColor, index) => {
+    if (typeof seriesColor === 'string') {
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:202]:
```suggestion
    if (isString(seriesColor)) {
```

File: packages/components/progress/src/progress.vue [hunk 4]
```diff
@@ -88,176 +88,146 @@ import {
...
     }
-  },
+    return seriesColor
+  })
+  return seriesColors.sort((a, b) => a.percentage - b.percentage)
+}
+
+const getCurrentColor = (percentage: number) => {
+  const { color } = props
+  if (typeof color === 'function') {
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:215]:
```suggestion
  if (isFunction(color)) {
```

File: packages/components/progress/src/progress.vue [hunk 5]
```diff
@@ -88,176 +88,146 @@ import {
...
+    return color(percentage)
+  } else if (typeof color === 'string') {
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:217]:
```suggestion
  } else if (isString(color)) {
```

**Review by jw-foss — COMMENTED** [review]
Left a few comments for you to consider 🏓 back to you

File: packages/components/progress/src/progress.vue (hunk 1 above)
Comment (chenxch) [reply packages/components/progress/src/progress.vue:91]: done.

File: packages/components/progress/src/progress.ts [hunk 6]
```diff
@@ -1,7 +1,8 @@
 import { buildProps, definePropType } from '@element-plus/utils'
 import type { ExtractPropTypes, SVGAttributes } from 'vue'
+import type Progress from './progress.vue'
 
-type Color = { color: string; percentage: number }
+export type Color = { color: string; percentage: number }
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:5]: Please add prefix `Progress`, `ProgressColor`

File: packages/components/progress/src/progress.ts [hunk 7]
```diff
@@ -1,7 +1,8 @@
...
 type ProgressFn = (percentage: number) => string
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:6]: export this as well

Comment (chenxch) [reply packages/components/progress/src/progress.ts]: processed

**Comment by chenxch** [comment]
@sxzz Thanks for your review, all processed.

File: packages/components/progress/src/progress.vue [hunk 8]
```diff
@@ -87,177 +87,148 @@ import {
...
-        width: `${props.percentage}%`,
-        animationDuration: `${props.duration}s`,
-        backgroundColor: getCurrentColor(props.percentage),
-      })
-    )
+})
 
-    const relativeStrokeWidth = computed(() =>
-      ((props.strokeWidth / props.width) * 100).toFixed(1)
+const COLOR_STATUS_MAP = {
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:99]:
```suggestion
const STATUS_COLOR_MAP = {
```

File: packages/components/progress/src/progress.vue [hunk 9]
```diff
@@ -87,177 +87,148 @@ import {
...
+  exception: '#ff4949',
+  warning: '#e6a23c',
+  default: '#20a0ff',
+}
+
+const props = defineProps(progressProps)
+
+const ns = useNamespace('progress')
+
+const barStyle = computed(
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:110]:
```suggestion
const barStyle = computed<CSSProperties>(
```

File: packages/components/progress/src/progress.vue [hunk 10]
```diff
@@ -87,177 +87,148 @@ import {
...
+
+const radius = computed(() => {
+  if (['circle', 'dashboard'].includes(props.type)) {
+    return Number.parseInt(
+      `${50 - Number.parseFloat(relativeStrokeWidth.value) / 2}`,
+      10
     )
+  } else {
+    return 0
+  }
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:130]:
```suggestion
  }
  return 0
```

File: packages/components/progress/src/progress.vue [hunk 11]
```diff
@@ -87,177 +87,148 @@ import {
...
+  return `${offset}px`
+})
 
-    const progressTextSize = computed(() => {
-      return props.type === 'line'
-        ? 12 + props.strokeWidth * 0.4
-        : props.width * 0.111111 + 2
-    })
+const trailPathStyle = computed(
+  (): CSSProperties => ({
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:154]:
```suggestion
const trailPathStyle = computed<CSSProperties>(
  () => ({
```

File: packages/components/progress/src/progress.vue [hunk 12]
```diff
@@ -87,177 +87,148 @@ import {
...
+    strokeDasharray: `${perimeter.value * rate.value}px, ${perimeter.value}px`,
+    strokeDashoffset: strokeDashoffset.value,
+  })
+)
+
+const circlePathStyle = computed(
+  (): CSSProperties => ({
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:161]:
```suggestion
const circlePathStyle = computed<CSSProperties>(
  () => ({
```

File: packages/components/progress/src/progress.vue [hunk 13]
```diff
@@ -87,177 +87,148 @@ import {
...
-    const content = computed(() => props.format(props.percentage))
+const statusIcon = computed(() => {
+  if (props.status === 'warning') {
+    return WarningFilled
+  }
+  if (props.type === 'line') {
+    return props.status === 'success' ? CircleCheck : CircleClose
+  } else {
+    return props.status === 'success' ? Check : Close
+  }
```
Comment (JeremyWuuuuu) [inline packages/components/progress/src/progress.vue:189]:
```suggestion
  switch(props.status) {
    case 'warning':
      return WarningFilled
    case 'line':
      return props.status === 'success' ? CircleCheck : CircleClose
    default:
      return props.status === 'success' ? Check : Close
  }
```

**Review by jw-foss — COMMENTED** [review]
@chenxch I left a few comments for you to consider, back to you 🏓

File: packages/components/progress/src/progress.vue (hunk 13 above)
Comment (chenxch) [reply packages/components/progress/src/progress.vue:189]: `line` is a type, if `const type = props.status === 'warning' ? 'warning' : props.type` This seems to be an extra step

**Comment by chenxch** [comment]
@JeremyWuuuuu Thanks for your careful review.🤟🏼

**Review by jw-foss — APPROVED** [review]

**Comment by JeremyWuuuuu** [comment]
@chenxch I need you to do this:

Since you have 5 commits.

```shell
$ git rebase -i HEAD~5
```

Replace the `pick` with `s` or `squash` for the top 4, don't touch the last one. After this is done, you can save and exit the editor.

Or you may handle this with GUI client like GitHub or SourceTree.

**Review by sxzz — APPROVED** [review]

**Pull request merged**